python-stdlib/logging/logging.py

```diff
@@ -1,128 +1,212 @@
+        self.log(INFO, message, *args)
 
-    def info(self, msg, *args):
-        self.log(INFO, msg, *args)
+    def warning(self, message, *args):
+        self.log(WARNING, message, *args)
 
-    def warning(self, msg, *args):
-        self.log(WARNING, msg, *args)
+    def error(self, message, *args):
+        self.log(ERROR, message, *args)
 
-    def error(self, msg, *args):
-        self.log(ERROR, msg, *args)
+    def critical(self, message, *args):
+        self.log(CRITICAL, message, *args)
 
-    def critical(self, msg, *args):
-        self.log(CRITICAL, msg, *args)
+    def exception(self, message, *args):
+        self.log(ERROR, message, *args)
+        if hasattr(sys, "exc_info"):
+            for h in filter(lambda h: isinstance(h, StreamHandler), self.handlers):
+                sys.print_exception(sys.exc_info()[1], h.stream)
 
-    def exc(self, e, msg, *args):
-        self.log(ERROR, msg, *args)
-        sys.print_exception(e, _stream)
+    def log(self, level, message, *args):
+        if self.isEnabledFor(level) and self.handlers:
+            if args and isinstance(args[0], dict):
+                args = args[0]
+            for h in self.handlers:
+                self.record.set(self.name, level, message % args)
+                h.emit(self.record)
 
-    def exception(self, msg, *args):
-        self.exc(sys.exc_info()[1], msg, *args)
 
-    def addHandler(self, hndlr):
-        self.handlers.append(hndlr)
+def debug(message, *args):
+    getLogger().log(DEBUG, message, *args)
 
 
-_level = INFO
-_loggers = {}
+def info(message, *args):
+    getLogger().log(INFO, message, *args)
 
 
-def getLogger(name="root"):
-    if name in _loggers:
-        return _loggers[name]
-    l = Logger(name)
-    _loggers[name] = l
-    return l
+def warning(message, *args):
+    getLogger().log(WARNING, message, *args)
 
 
-def info(msg, *args):
-    getLogger().info(msg, *args)
+def error(message, *args):
+    getLogger().log(ERROR, message, *args)
 
 
-def debug(msg, *args):
-    getLogger().debug(msg, *args)
+def critical(message, *args):
+    getLogger().log(CRITICAL, message, *args)
 
 
-def basicConfig(level=INFO, filename=None, stream=None, format=None):
-    global _level, _stream
-    _level = level
-    if stream:
-        _stream = stream
-    if filename is not None:
-        print("logging.basicConfig: filename arg is not supported")
-    if format is not None:
-        print("logging.basicConfig: format arg is not supported")
+def exception(message, *args):
+    getLogger().exception(message, *args)
+
+
+def shutdown():
+    for k, logger in loggers.items():
+        for h in logger.handlers:
+            h.close()
+        loggers.pop(logger, None)
+
+
+def getLogger(name="root"):
+    if name not in loggers:
+        loggers[name] = Logger(name)
+    return loggers[name]
+
+
+def basicConfig(
+    filename=None,
+    filemode="a",
+    format=default_fmt,
+    datefmt=default_datefmt,
+    level=WARNING,
```

Comment:
the existing default was `INFO`, but I think `WARNING` matches CPython, so probably a good time to change this default